### PR TITLE
Mod3

### PR DESCRIPTION
**Pull request by kcny**

### CODE REVIEW

File: app/controllers/forgot_passwords_controller.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    if user
+      AppMailer.send_forgot_password(user).deliver
+      redirect_to forgot_password_confirmation_path
+    else
+      flash[:error] = params[:email].blank? ? "Email cannot be blank." : "User email does not exist."
+      redirect_to forgot_password_path
+    end
+  end
+
+  def confirm
```
Comment (tomtomecek) [inline app/controllers/forgot_passwords_controller.rb]: I would suggest to use `def confirm; end` or just do not use it. But the former looks better :).

File: app/controllers/forgot_passwords_controller.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class ForgotPasswordsController < ApplicationController
+
+  def create
+    user = User.where(email: params[:email]).first
+    if user
+      AppMailer.send_forgot_password(user).deliver
+      redirect_to forgot_password_confirmation_path
+    else
+      flash[:error] = params[:email].blank? ? "Email cannot be blank." : "User email does not exist."
```
Comment (alagram) [inline app/controllers/forgot_passwords_controller.rb]: You should use `flash[:danger]` here to match current bootstrap classes. That said, I've seen some fixes which handle this transformation so if you have that, no problem.

File: app/controllers/forgot_passwords_controller.rb (hunk 1 above)
Comment (alagram) [inline app/controllers/forgot_passwords_controller.rb]: You can remove this action if it has no logic and rails will automatically render the the confirm template.

File: app/views/forgot_passwords/new.html.haml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    .col-sm-10.col-bg-offset-1
+      = form_tag forgot_passwords_path, class: "login" do 
+        %header
+          %h1 Forgot Password?
+          %p We will send you an email with a link that you can use to reset your password.
+        .form-group
+          = label_tag :email, "Email Address"
+          .row
+            .col-sm-4
+              = text_field_tag :email
```
Comment (alagram) [inline app/views/forgot_passwords/new.html.haml]: I believe you can also use `email_field_tag` here. Rails will change the type of the input to 'email'.

File: app/views/password_resets/show.html.haml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  .row
+    .col-sm-10.col-sm-offset-1
+      = form_tag password_resets_path do 
+        %header
+          %h1 Reset Your Password
+        %fieldset.form-group
+          = label_tag :password, "New Password"
+          .row
+            .col-sm-4
+              = text_field_tag :password
```
Comment (alagram) [inline app/views/password_resets/show.html.haml]: You can also use `password_field_tag` here to obfuscate the user's input.

File: spec/controllers/forgot_passwords_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        expect(response).to redirect_to forgot_password_path
+      end
+      it "shows an error message" do 
+        post :create, email: ''
+        expect(flash[:error]).to eq("Email cannot be blank.")
+      end
+    end
+    context "with existing email" do 
+      it "should redirect to a forgot password information page" do
+      Fabricate(:user, email: "hoza@example.com")
```
Comment (alagram) [inline spec/controllers/forgot_passwords_controller_spec.rb]: Lines 17-19 should be indented by two spaces.

File: spec/controllers/forgot_passwords_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      end
+    end
+    context "with existing email" do 
+      it "should redirect to a forgot password information page" do
+      Fabricate(:user, email: "hoza@example.com")  
+      post :create, email: "hoza@example.com"
+      expect(response).to redirect_to forgot_password_confirmation_path 
+      end 
+      it "sends out an email to user's email address." do
+      Fabricate(:user, email: "hoza@example.com")
```
Comment (alagram) [inline spec/controllers/forgot_passwords_controller_spec.rb]: Lines 22 - 26 as well.

File: app/controllers/forgot_passwords_controller.rb (hunk 1 above)
Comment (kcny) [inline app/controllers/forgot_passwords_controller.rb]: Done

File: app/controllers/forgot_passwords_controller.rb (hunk 2 above)
Comment (kcny) [inline app/controllers/forgot_passwords_controller.rb]: Done.

File: app/views/forgot_passwords/new.html.haml (hunk 3 above)
Comment (kcny) [inline app/views/forgot_passwords/new.html.haml]: Done.

File: app/views/password_resets/show.html.haml (hunk 4 above)
Comment (kcny) [inline app/views/password_resets/show.html.haml]: Done.

File: spec/controllers/forgot_passwords_controller_spec.rb (hunk 5 above)
Comment (kcny) [inline spec/controllers/forgot_passwords_controller_spec.rb]: Done.

File: spec/controllers/forgot_passwords_controller_spec.rb (hunk 6 above)
Comment (kcny) [inline spec/controllers/forgot_passwords_controller_spec.rb]: Done.

**Pull request merged**